src/suspendplugin.c

```diff
@@ -407,19 +460,16 @@ suspend_plugin_init()
         mce_display_add_state_changed_handler(mce_display,
             suspend_display_event, NULL);
 
-    nl_socket = nl_socket_alloc();
+    nl_socket_fd = nl_socket_get_fd(nl_socket);
+    netlink_channel = g_io_channel_unix_new(nl_socket_fd);
+    g_io_channel_set_close_on_unref(netlink_channel, TRUE);
 
-    genl_connect(nl_socket);
+    g_io_channel_set_encoding(netlink_channel, NULL, NULL);
+    g_io_channel_set_buffered(netlink_channel, FALSE);
 
-    driver_id = genl_ctrl_resolve(nl_socket, "nl80211");
-
-    ret = nl_get_multicast_id(nl_socket, "nl80211", "testmode");
-    if (ret >= 0) {
-        ret = nl_socket_add_membership(nl_socket, ret);
-    } else {
-        connman_error("Failed to register for testmode multicast group.\n");
-        return -1;
-    }
+    g_io_add_watch(netlink_channel,
+                G_IO_IN | G_IO_NVAL | G_IO_HUP | G_IO_ERR,
+                        suspend_plugin_netlink_handler, NULL);
```

Comment:
Unless I'm totally mistaken you need to store source id returned from g_io_add_watch() - otherwise the channel has extra ref until the callback returns false -> g_io_channel_unref() on exit path does not actually do anything.

I'd change this so that you keep the source id in the books, unref the channel before returning from here & do g_source_remove() at exit instead of  g_io_channel_unref(). In which case every "return false" from suspend_plugin_netlink_handler() should also clear the cached source id as it gets invalidated.

Reply:
Yes, the source id returned by `g_io_add_watch()` shouldn't be discarded.